File: enterprise_auth/api/sdk/python/exceptions.py
"""
EnterpriseAuth Python SDK Exceptions

Custom exception classes for the EnterpriseAuth SDK.
"""
from typing import Dict, Any, Optional


class EnterpriseAuthError(Exception):
    """Base exception for EnterpriseAuth SDK errors."""
    
    def __init__(self, message: str, details: Optional[Dict[str, Any]] = None):
        super().__init__(message)
        self.message = message
        self.details = details or {}


class AuthenticationError(EnterpriseAuthError):
    """Raised when authentication fails."""
    pass


class AuthorizationError(EnterpriseAuthError):
    """Raised when authorization fails (insufficient permissions)."""
    pass


class RateLimitError(EnterpriseAuthError):
    """Raised when rate limit is exceeded."""
    
    def __init__(self, message: str, retry_after: int = 60, details: Optional[Dict[str, Any]] = None):
        super().__init__(message, details)
        self.retry_after = retry_after


class ValidationError(EnterpriseAuthError):
    """Raised when request validation fails."""
    
    def __init__(self, message: str, details: Optional[Dict[str, Any]] = None):
        super().__init__(message, details)
        self.validation_errors = details or {}


class WebhookError(EnterpriseAuthError):
    """Raised when webhook operations fail."""
    pass


class NetworkError(EnterpriseAuthError):
    """Raised when network operations fail."""
    pass


class TimeoutError(EnterpriseAuthError):
    """Raised when requests timeout."""
    pass

File: enterprise_auth/api/sdk/javascript/src/exceptions.ts
/**
 * EnterpriseAuth SDK Exceptions
 *
 * Custom exception classes for the JavaScript SDK.
 */

export class EnterpriseAuthError extends Error {
  public code?: string;
  public details?: Record<string, any>;

  constructor(message: string, code?: string, details?: Record<string, any>) {
    super(message);
    this.name = "EnterpriseAuthError";
    this.code = code;
    this.details = details;

    // Maintains proper stack trace for where our error was thrown (only available on V8)
    if (Error.captureStackTrace) {
      Error.captureStackTrace(this, EnterpriseAuthError);
    }
  }

  toString(): string {
    if (this.code) {
      return `[${this.code}] ${this.message}`;
    }
    return this.message;
  }

  toJSON(): Record<string, any> {
    return {
      name: this.name,
      message: this.message,
      code: this.code,
      details: this.details,
      stack: this.stack,
    };
  }
}

export class AuthenticationError extends EnterpriseAuthError {
  constructor(message: string, code?: string, details?: Record<string, any>) {
    super(message, code, details);
    this.name = "AuthenticationError";
  }
}

export class AuthorizationError extends EnterpriseAuthError {
  constructor(message: string, code?: string, details?: Record<string, any>) {
    super(message, code, details);
    this.name = "AuthorizationError";
  }
}

export class RateLimitError extends EnterpriseAuthError {
  public retryAfter?: number;

  constructor(
    message: string,
    retryAfter?: number,
    code?: string,
    details?: Record<string, any>
  ) {
    super(message, code, details);
    this.name = "RateLimitError";
    this.retryAfter = retryAfter;
  }

  toJSON(): Record<string, any> {
    return {
      ...super.toJSON(),
      retryAfter: this.retryAfter,
    };
  }
}

export class ValidationError extends EnterpriseAuthError {
  public fieldErrors?: Record<string, string[]>;

  constructor(
    message: string,
    fieldErrors?: Record<string, string[]>,
    code?: string,
    details?: Record<string, any>
  ) {
    super(message, code, details);
    this.name = "ValidationError";
    this.fieldErrors = fieldErrors;
  }

  toJSON(): Record<string, any> {
    return {
      ...super.toJSON(),
      fieldErrors: this.fieldErrors,
    };
  }
}

export class WebhookError extends EnterpriseAuthError {
  constructor(message: string, code?: string, details?: Record<string, any>) {
    super(message, code, details);
    this.name = "WebhookError";
  }
}

export class TokenError extends EnterpriseAuthError {
  constructor(message: string, code?: string, details?: Record<string, any>) {
    super(message, code, details);
    this.name = "TokenError";
  }
}

export class NetworkError extends EnterpriseAuthError {
  constructor(message: string, code?: string, details?: Record<string, any>) {
    super(message, code, details);
    this.name = "NetworkError";
  }
}

export class ConfigurationError extends EnterpriseAuthError {
  constructor(message: string, code?: string, details?: Record<string, any>) {
    super(message, code, details);
    this.name = "ConfigurationError";
  }
}
